File: src/patterns/source.ts
import type { Resource } from "./resource.js"
import type { GlobRule, Rule, RuleMatch } from "./rule.js"

/**
 * Represents a source of external patterns.
 *
 * @since 0.6.0
 */
export type Source = {
	/**
	 * Parent source for hierarchical ignore file rules.
	 *
	 * @since 0.12.0
	 */
	parent?: Resource
	/**
	 * @internal
	 *
	 * @since 0.11.0
	 */
	_noMatchCache?: RuleMatch
	/**
	 * Patterns defined within the source file.
	 * Those patterns are for ignoring files.
	 *
	 * @see {@link ruleTest}
	 *
	 * @since 0.11.0
	 */
	rules: Rule[]

	/**
	 * Relative path to the source file.
	 *
	 * @since 0.6.0
	 */
	path: string

	/**
	 * Indicates if the matching logic is inverted.
	 * For example, `package.json` `files` field inverts the matching logic,
	 * because it specifies files to include rather than exclude.
	 *
	 * @see {@link ruleTest}
	 *
	 * @since 0.6.0
	 */
	inverted: boolean

	/**
	 * Directory where the source was located.
	 *
	 * @since 0.12.0
	 */
	dir?: string
}

/**
 * Converts pattern ("x" (excludes) or "!x" (includes)) to a rule.
 * You can also invert the behavior.
 * It compiles the rule.
 *
 *
 * if !x -> includes + x
 * if x -> excludes + x
 * if invert && !x -> excludes + x
 * if invert && x -> includes + x
 *
 * @since 0.6.0
 */
export function resolveNegatable(pattern: string, invert: boolean, reuse?: GlobRule): GlobRule {
	const isEscapedBang =
		pattern.length > 1 && pattern.charCodeAt(0) === 92 && pattern.charCodeAt(1) === 33
	let negated = false
	if (isEscapedBang) {
		pattern = pattern.slice(1)
	} else if (pattern.charCodeAt(0) === 33) {
		negated = true
		pattern = pattern.slice(1)
	}
	const excludes = negated === invert
	const iff = reuse && excludes === reuse.excludes
	const rule: GlobRule = iff ? reuse : { compiled: null, excludes, list: [] }
	rule.list.push(pattern)
	return rule
}

File: internal/shared/source.go
package shared

// Represents a source of patterns for matching paths.
//
// # Since 0.6.0
type Source struct {
	// Patterns defined within the source file.
	// Those patterns are for ignoring files.
	//
	// # Since 0.6.0
	Pattern []SignedPattern

	// Name of the source file.
	//
	// # Since 0.6.0
	Name string

	// Relative path to the source file.
	//
	// # Since 0.6.0
	Path string

	// Indicates if the matching logic is inverted.
	// For example, `package.json` `files` field inverts the matching logic,
	// because it specifies files to include rather than exclude.
	//
	// # Since 0.6.0
	Inverted bool

	// Error encountered during extraction, if any.
	//
	// See [SourceExtractor].
	//
	// # Since 0.6.0
	Error error
}

// Adds a negatable pattern to the source's pattern lists.
// Strips the leading '!' for include patterns,
// and adds to exclude patterns otherwise.
//
// # Since 0.6.0
func (source *Source) PushNegatable(
	pattern string,
	invert bool,
	include, exclude SignedPattern,
) {
	if invert {
		exclude, include = include, exclude
	}

	dist := exclude

	if len(pattern) > 0 && pattern[0] == '!' {
		dist = include
		pattern = pattern[1:]
	}

	dist.Pattern = append(dist.Pattern, pattern)
}
